src/util/Generators.h

```diff
@@ -22,30 +24,67 @@ namespace ad_utility {
 // returns false. If the `aggregator` returns false, the cached value is
 // discarded. If the cached value is still present once the generator is fully
 // consumed, `onFullyCached` is called with the cached value.
-// NOTE: The `int` is just a dummy value.
+template <typename InputRange, typename AggregatorT, typename FullyCachedFuncT>
+class CachingWrapper
+    : public InputRangeFromGet<ql::ranges::range_value_t<InputRange>> {
+ public:
+  using T = ql::ranges::range_value_t<InputRange>;
+
+  CachingWrapper(InputRange generator, AggregatorT aggregator,
+                 FullyCachedFuncT onFullyCached)
+      : generator_(std::move(generator)),
+        aggregator_(std::move(aggregator)),
+        onFullyCached_(std::move(onFullyCached)) {}
```

Comment:
Make the members public (you hide the whole struct anyway), and delete the constructor and then use aggregate initialization (less boilerplate)